File: solutions/julia/roman-numerals/1/roman-numerals.jl
function to_roman(number)
    result = ""
    letters = ["M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"]
    numbers = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1]

    if number < 1
        error("Number must be 1 or greater!")
    end
    
    for i in eachindex(numbers)
        while number >= numbers[i]
            result = result * letters[i]
            number = number - numbers[i]
        end
    end
    result
end

File: solutions/typescript/roman-numerals/1/roman-numerals.ts
export const toRoman = (arabic: number): string => {
  const DIVISORS: number[] = [1000, 100, 10, 1];

  const DICTIONARY: {
    [d: number]: { base: string; five: string; next: string };
  } = {
    1000: {
      base: 'M',
      five: '',
      next: '',
    },
    100: {
      base: 'C',
      five: 'D',
      next: 'M',
    },
    10: {
      base: 'X',
      five: 'L',
      next: 'C',
    },
    1: {
      base: 'I',
      five: 'V',
      next: 'X',
    },
  };

  let result: string = '';

  for (const divisor of DIVISORS) {
    if (arabic >= divisor) {
      const times = Math.floor(arabic / divisor);
      const { base, five, next } = DICTIONARY[divisor];
      if (times === 9) result += base + next;
      else if (times >= 5) result += five + base.repeat(times - 5);
      else if (times === 4) result += base + five;
      else result += base.repeat(times);
      arabic %= divisor * times;
    }
  }

  return result;
};
